node/src/tasks/msp_move_bucket.rs

```diff
@@ -0,0 +1,306 @@
+                .await?;
+
+            return Ok(());
+        };
+
+        let call = storage_hub_runtime::RuntimeCall::FileSystem(
+            pallet_file_system::Call::msp_respond_move_bucket_request {
+                bucket_id: event.bucket_id,
+                response: BucketMoveRequestResponse::Accepted,
+            },
+        );
+
+        info!(
+            target: LOG_TARGET,
+            "MSP: accepting move bucket request for bucket {:?}",
+            event.bucket_id,
+        );
+
+        self.storage_hub_handler
+            .blockchain
+            .send_extrinsic(call, Tip::from(0))
+            .await?
+            .with_timeout(Duration::from_secs(
+                self.storage_hub_handler
+                    .provider_config
+                    .extrinsic_retry_timeout,
+            ))
+            .watch_for_success(&self.storage_hub_handler.blockchain)
+            .await?;
+
+        let bucket = event.bucket_id.as_ref().to_vec();
+
+        let mut indexer_connection = indexer_db_pool.get().await?;
+
+        let forest_storage = self
+            .storage_hub_handler
+            .forest_storage_handler
+            .get_or_create(&event.bucket_id.as_ref().to_vec())
+            .await;
+
+        for file in shc_indexer_db::models::File::get_by_onchain_bucket_id(
+            &mut indexer_connection,
+            bucket.clone(),
+        )
+        .await?
+        {
+            info!(
+                target: LOG_TARGET,
+                "MSP: downloading file {:?} of bucket {:?}",
+                file.file_key,
+                event.bucket_id,
+            );
+
+            let file_metadata = file.to_file_metadata(bucket.clone());
+            let file_key = file_metadata.file_key::<HashT<StorageProofsMerkleTrieLayout>>();
+
+            self.storage_hub_handler
+                .file_storage
+                .write()
+                .await
+                .insert_file(file_key.clone(), file_metadata.clone())
+                .unwrap();
+
+            let chunks_count = file_metadata.chunks_count();
+
+            let mut bsp_peer_ids = file.get_bsp_peer_ids(&mut indexer_connection).await?;
+
+            // Shuffle in order to avoid consecutive requests to the same BSP node.
+            bsp_peer_ids.shuffle(&mut rand::thread_rng());
+
+            if bsp_peer_ids.is_empty() {
+                error!(
+                    target: LOG_TARGET,
+                    "No BSP peer IDs found for file {:?} of bucket {:?}",
+                    file_key, event.bucket_id,
+                );
+                continue;
+            }
+
+            // We will cycle through all the BSP peer IDs for each chunk until we successfully
+            // download the file.
+            let mut bsp_peer_ids_iter = bsp_peer_ids.iter().cycle();
+
+            for chunk in 0..chunks_count {
+                for _ in 0..DOWNLOAD_REQUEST_RETRY_COUNT {
+                    let peer_id = bsp_peer_ids_iter.next().unwrap();
```

Comment:
Shouldn't we cycle on files instead of chunks. I think this would be more efficient to request all the chunks from a single BSP.

Reply:
I was thinking more about distributing the load to the BSPs than getting the file as fast as possible. Anyway - the fastest transfer speeds would be achieved if we parallelize this (have multiple inflight download requests at the same time). This way you can actually saturate the download bandwidth with multiple uploaders.